google-cloud-env/lib/google/cloud/env.rb

```diff
@@ -273,26 +275,36 @@ def app_engine_memory_mb
       end
 
       ##
-      # Returns the name of the Container Engine cluster hosting the
+      # Returns the name of the Kubernetes Engine cluster hosting the
       # application, or `nil` if the current code is not running in
-      # Container Engine.
+      # Kubernetes Engine.
       #
       # @return [String,nil]
       #
-      def container_engine_cluster_name
+      def kubernetes_engine_cluster_name
         instance_attribute "cluster-name"
       end
+      alias container_engine_cluster_name kubernetes_engine_cluster_name
 
       ##
-      # Returns the name of the Container Engine namespace hosting the
+      # Returns the name of the Kubernetes Engine namespace hosting the
       # application, or `nil` if the current code is not running in
-      # Container Engine.
+      # Kubernetes Engine.
       #
       # @return [String,nil]
       #
-      def container_engine_namespace_id
-        env["GKE_NAMESPACE_ID"]
+      def kubernetes_engine_namespace_id
+        # The Kubernetes namespace is difficult to obtain without help from
+        # the application using the Downward API. The environment variable
+        # below is set in some older versions of GKE, and the file below is
+        # present in Kubernetes as of version 1.9, but it is possible that
+        # alternatives will need to be found in the future.
+        env["GKE_NAMESPACE_ID"] ||
+          ::IO.read("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
```

Comment:
Why `IO.read` and not `File.read`?

Reply:
I guess I've always used `IO.read` because the method is actually defined on `IO` rather than on `File`. If you think it's more clear to use `File`, I have no objections to changing it.